### ISSUE TITLE
[XF] Add a dialog service

### ISSUE THREAD

**Issue by brianlagunas** [PrismLibrary/Prism#76]
Add a service that will allow a developer to call Page.DisplayAlert from within a ViewModel.

**Comment by noufionline** [comment]
Thanks a lot Brian

**Comment by YehudahA** [comment]
Suggestion how to implement it:

##### IDialogViewModel
``` cs
public interface IDialogViewModel
{
    string Title { get; }
    event EventHandler Closed;
}
```

#####  IDialogService
``` cs
public interface IDialogService
{
    void ShowDialog<T>() where T : IDialogViewModel;
    void ShowDialog<T>(T context) where T : IDialogViewModel;
    void ShowDialog<T>(T context, Action<T> callback) where T : IDialogViewModel;
}
```
##### DialogService
``` cs
[Export(typeof(IDialogService))]
public sealed class DialogService : IDialogService
{
    #region fields - ctor

    readonly IServiceLocator serviceLocator;

    [ImportingConstructor]
    public DialogService(IServiceLocator serviceLocator)
    {
        this.serviceLocator = serviceLocator;
    }

    #endregion // fields - ctor

    #region IDialogService

    public void ShowDialog<T>() where T : IDialogViewModel
    {
        IDialogViewModel context = serviceLocator.GetInstance(typeof(T)) as IDialogViewModel;
        ShowDialog(context);
    }

    public void ShowDialog<T>(T context) where T : IDialogViewModel
    {
        Invoke(context, null);
    }

    public void ShowDialog<T>(T context, Action<T> callback) where T : IDialogViewModel
    {
        Invoke(context, callback);
    }

    #endregion // IDialogService

    #region Invoke

    private void Invoke<T>(T dialogViewModel, Action<T> callback) where T : IDialogViewModel
    {
        Window wrapperWindow = GetWindow(dialogViewModel);

        EventHandler viewModelClosedHandler = (sender, e) => wrapperWindow.Close();
        dialogViewModel.Closed += viewModelClosedHandler;

        // We invoke the callback when the interaction's window is closed.
        EventHandler windowClosedHandler = null;

        windowClosedHandler =
            (sender, e) =>
            {
                wrapperWindow.Closed -= windowClosedHandler;
                dialogViewModel.Closed -= viewModelClosedHandler;

                wrapperWindow.Content = null;
                if (callback != null) callback(dialogViewModel);
            };

        wrapperWindow.Closed += windowClosedHandler;
        wrapperWindow.ShowDialog();
    }

    private static Window GetWindow(IDialogViewModel dialogViewModel)
    {
        Window wrapperWindow = new Window
        {
            Title = dialogViewModel.Title,
            ShowInTaskbar = false,
            WindowStartupLocation = WindowStartupLocation.CenterOwner,
            WindowState = WindowState.Normal,
            SizeToContent = SizeToContent.WidthAndHeight
        };

        Window mainWindow = System.Windows.Application.Current.MainWindow;

        if (mainWindow != null)
        {
            wrapperWindow.Owner = mainWindow;
            wrapperWindow.Icon = mainWindow.Icon;
        }

        FrameworkElement view = YA.Windows.ViewLocationProvider.GetViewByViewModel(dialogViewModel);
        wrapperWindow.Content = view;

        return wrapperWindow;
    }

    #endregion // Invoke
}
```

**Comment by brianlagunas** [comment]
Thanks for the suggestion @YehudahA but this issues is not for WPF, but rather Xamarin.Forms.  This is why there is a "[XF]" in front of the issue.

**Comment by noufionline** [comment]
Hi Brian,

How about adding a dialogue service in wpf?




Regards,

Noufal



—

**Comment by brianlagunas** [comment]
@noufionline No, we will not be adding a dialog service to WPF.  Dialog services are usually very specific to an application and/or scenario that usually involve a number of customized needs.  It is so simple to create a dialog service, there is no need for us to ship one that will most likely not meet the needs of most devs.

**Comment by YehudahA** [comment]
:(

**Comment by joacar** [comment]
Hi,

First of I would introduce the `IPageResolver` interface with the single method `Page Resolve()` or similar. This way MasterDetailPage scenarios and others can be implemented in the application since they handle CurrentPage a bit differently.

    public class DefaultPageResolver : IPageResolver {
       public Page Resolve() { return Application.Current.MainPage; }
    }

    public int delegate DisplayActionSheetButtonDelegate()    

    public interface IDialogService {
     Task DisplayAlert(title, message,ok)
     Task<bool> DisplayAlert(title,message,ok,cancel)
     Task<int> DisplayActionSheet(string title, string cancel, string destruction, 
      params DisplayActionSheetButtonDelegate[] buttons); // Not sure if this works :)
     }

An implementation would be

    public class DialogService : IDialogService {
     private IPageResolver _pageResolver;      
     public DialogService(IPageResolver pageResolver) { _pageResolver = pageResolver; }
     
    #region Implementation of IDialogService

     public async Task DisplayAlert() { return away _pageResolver.Resolve().DisplayAlert(); }

     #endregion

    }

Wrote this just here so sorry for ugly code/format. 

I would be happy to write the actual code if requested and the above pattern is considered good enough. 

Later on adding support for displaying progress and similar..

**Comment by brianlagunas** [comment]
I'm curious how the proper PageResolver would be obtained when using different types of pages in a navigation lifecycle.  For example, let's say the MainPage of the app is a ContentPage upon loading.  We ask for the IDialogService in the Page's ViewModel, and everything works fine.  It will use the DefaultPageResolver.  Then we navigate away to a MasterDetailPage.  We again ask for the IDialogService in the constructor of the MasterDetailPage's ViewModel, but now how do we change out the resolver?  Since we are using DI to create our dependencies, we aren't passing them in manually.

**Comment by joacar** [comment]
Haven't stumbled into that scenario so I have not thought of it to be honest. 

The none-solution is to leave that implementation to the application and perform necessary checking to resolve the correct page depending on the type of MainPage currently active. 

If above is needed at all. Perhaps it "magically" just works as long as it is a Page reference.

Made a small test in my current project and dialog shows correctly on both button presses. Will invenstigate more during weekend, a bit short on time at the moment.

````c#

public class PageResolver : IPageResolver {
    Page Resolve() { return Application.Current.MainPage; }
}

var masterDetail = new MasterDetailPage
                {
                    Title = "MasterDetail",
                    Master = new ContentPage()
                    {
                        Title = "Master in detail",
                        Content = new StackLayout()
                        {
                            Children =
                            {
                                new Button()
                                {
                                    Text = "Hello from master",
                                    Command = new Command(() =>
                                    {
                                        var ds = App.Container.Resolve<IDialogService>();
                                        ds.DisplayAlert("Master", "Hello from master", "Cancel");
                                    })
                                }
                            }
                        }
                    },
                    Detail = new NavigationPage(new ContentPage()
                    {
                        Title = "Detail",
                        Content = new StackLayout()
                        {
                            Children =
                            {
                                new Button()
                                {
                                    Text = "Hello from detail",
                                    Command = new Command(() =>
                                    {
                                        var ds = App.Container.Resolve<IDialogService>();
                                        ds.DisplayAlert("Detail", "Hello from detail", "Cancel");
                                    })
                                }
                            }
                        }
                    })
                };
                Application.Current.MainPage = masterDetail;
````

**Comment by brianlagunas** [comment]
Good point.  The responsibility could probably just be placed on the app developer for their specific scenario.

Another thought, should this be called an AlertService instead?  My gut is telling me it should, but wanted to get your thoughts.

**Comment by joacar** [comment]
In my opionion DisplaySheet doesn't fall into the "alert" category as well as the others.

Perhaps we should introduce an `IDialogFacade` that wraps interfaces such as `IAlertService`, `IUserInputService`, `IProgressDialogService` or similar.

The contract for `IDialogFacade` could be

````c#
public interface IDialogFacade {
 Task DisplayAlertAsync();
Task<bool> DisplayAlertAsync();
Task<bool> DisplayAsync({Success, Alert,Warning});
Task<IProgressDialog> DisplayProgress();
}
````

These might be better patterna since view models doesn't have to request many interfaces for dealing with relatively common ways to interact/inform user about events.

Throws what i think might be an issue with navigation as well. When navigating with modal option, the page displayed modally, in android and ios, doesn't present the navigation bar unless explicility wrapped in a new `NavigationPage`. I will make a pure sample project to be 100% sure this is the case (I have noticed it in my current project) and when confirmed I will probably open an issue and suggest a way to push modal pages with an option to show navigation bar.

**Comment by brianlagunas** [comment]
This has been implemented by @joacar with PR #113
